In [1]:
import pandas as pd

In [5]:
matches=pd.read_csv("matches.csv", index_col=0)

In [10]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [11]:
matches.shape

(1389, 27)

### About EPL

### There are 38 matches played each season and there are 20 teams and we have data for 2 seasopns 2021 and 2022

### So there must be 38x20x2 =1520 rows but we have 1389 rows only, Thus we have some missing data

In [14]:
 matches["team"].value_counts()

team
Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: count, dtype: int64

In [17]:
matches["round"].value_counts()

round
Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: count, dtype: int64

## Cleaning the data

In [18]:
matches.dtypes

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [19]:
matches["date"]=pd.to_datetime(matches["date"])

In [20]:
matches.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

## Creating predictors for ML

In [21]:
## Venue_Code 0(Away Game) and 1(Home Game)
matches["venue_code"] = matches["venue"].astype("category").cat.codes
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  notes    sh   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...    NaN  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...    NaN  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...    NaN  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...    NaN  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...    NaN  16.0   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...    ...   ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...    NaN   8.0   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...    NaN   7.0   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...    NaN  10.0   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...    NaN  11.0   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...    NaN  12.0   3.0  17.0  0.0  0.0   0.0   2021   

                team  venue_code  
1    Manchester City           0  
2    Manchester City           1  
3    Manchester City           1  
4    Manchester City           0  
6    Manchester City           1  
..               ...         ...  
38  Sheffield United           0  
39  Sheffield United           1  
40  Sheffield United           0  
41  Sheffield United           0  
42  Sheffield United           1  

[1389 rows x 28 columns]

In [22]:
## Creating a unique code for each opponent
matches["opp_code"]=matches["opponent"].astype("category").cat.codes
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...    sh   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...  16.0   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...   ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...   8.0   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...   7.0   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...  10.0   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...  11.0   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...  12.0   3.0  17.0  0.0  0.0   0.0   2021   

                team  venue_code  opp_code  
1    Manchester City           0        18  
2    Manchester City           1        15  
3    Manchester City           1         0  
4    Manchester City           0        10  
6    Manchester City           1        17  
..               ...         ...       ...  
38  Sheffield United           0        18  
39  Sheffield United           1         6  
40  Sheffield United           0         7  
41  Sheffield United           0        14  
42  Sheffield United           1         4  

[1389 rows x 29 columns]

In [26]:
 ## Creating an hour column for the time(only in hour) on which the matches are played
matches["hour"]=matches["time"].str.replace(":.+", "", regex=True).astype("int")
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...   3.0  17.0  0.0  0.0   0.0   2021   

                team venue_code  opp_code  hour  
1    Manchester City          0        18    16  
2    Manchester City          1        15    15  
3    Manchester City          1         0    12  
4    Manchester City          0        10    15  
6    Manchester City          1        17    15  
..               ...        ...       ...   ...  
38  Sheffield United          0        18    19  
39  Sheffield United          1         6    15  
40  Sheffield United          0         7    19  
41  Sheffield United          0        14    18  
42  Sheffield United          1         4    16  

[1389 rows x 30 columns]

In [27]:
## Creating weekcode i.e a unique code for each day of the week

In [30]:
matches["day_code"]=matches["date"].dt.dayofweek
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   fk   pk  pkatt  season              team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022   Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022   Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022   Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022   Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022   Manchester City   
..  ...             ...  ...  ...  ...    ...     ...               ...   
38  4.0       Tottenham  ...  0.0  0.0    0.0    2021  Sheffield United   
39  2.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Sheffield United   
40  0.0         Everton  ...  0.0  0.0    0.0    2021  Sheffield United   
41  1.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Sheffield United   
42  0.0         Burnley  ...  0.0  0.0    0.0    2021  Sheffield United   

   venue_code opp_code hour  dayt_code  day_code  
1           0       18   16          6         6  
2           1       15   15          5         5  
3           1        0   12          5         5  
4           0       10   15          5         5  
6           1       17   15          5         5  
..        ...      ...  ...        ...       ...  
38          0       18   19          6         6  
39          1        6   15          5         5  
40          0        7   19          6         6  
41          0       14   18          2         2  
42          1        4   16          6         6  

[1389 rows x 32 columns]

In [36]:
## ignore this|
matches.drop("dayt_code",axis=1)

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  dist   fk   pk  pkatt season              team  \
1   1.0       Tottenham  ...  16.9  1.0  0.0    0.0   2022   Manchester City   
2   0.0    Norwich City  ...  17.3  1.0  0.0    0.0   2022   Manchester City   
3   0.0         Arsenal  ...  14.3  0.0  0.0    0.0   2022   Manchester City   
4   0.0  Leicester City  ...  14.0  0.0  0.0    0.0   2022   Manchester City   
6   0.0     Southampton  ...  15.7  1.0  0.0    0.0   2022   Manchester City   
..  ...             ...  ...   ...  ...  ...    ...    ...               ...   
38  4.0       Tottenham  ...  17.4  0.0  0.0    0.0   2021  Sheffield United   
39  2.0  Crystal Palace  ...  11.4  1.0  0.0    0.0   2021  Sheffield United   
40  0.0         Everton  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   
41  1.0   Newcastle Utd  ...  16.0  1.0  0.0    0.0   2021  Sheffield United   
42  0.0         Burnley  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   

   venue_code opp_code  hour  day_code  
1           0       18    16         6  
2           1       15    15         5  
3           1        0    12         5  
4           0       10    15         5  
6           1       17    15         5  
..        ...      ...   ...       ...  
38          0       18    19         6  
39          1        6    15         5  
40          0        7    19         6  
41          0       14    18         2  
42          1        4    16         6  

[1389 rows x 31 columns]